devices/src/ioapic.rs

```diff
@@ -0,0 +1,339 @@
+// Copyright 2019 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+//
+// Implementation of an intel 82093AA Input/Output Advanced Programmable Interrupt Controller
+// See https://pdos.csail.mit.edu/6.828/2016/readings/ia32/ioapic.pdf for a specification.
+
+use crate::BusDevice;
+use byteorder::{ByteOrder, LittleEndian};
+use std::{io, result};
+
+#[derive(Debug)]
+pub enum Error {
+    /// Interrupt has been blocked by the guest.
+    InterruptGuestBlocked,
+    /// Failed to send an interrupt.
+    InterruptFailed(io::Error),
+    /// Invalid destination mode.
+    InvalidDestinationMode,
+    /// Invalid trigger mode.
+    InvalidTriggerMode,
+    /// Invalid delivery mode.
+    InvalidDeliveryMode,
+}
+
+type Result<T> = result::Result<T, Error>;
+
+pub type IoapicInterrupt = Box<Fn(MsiMessage) -> result::Result<(), Error> + Send>;
```

Comment:
Actually, why do we need this to be dynamic? When will we need anything but what's defined in
https://github.com/intel/cloud-hypervisor/pull/57/commits/4b2d4fe17b9b3084b63640d8078a642b9c63e3a1#diff-a375fb00f71d40c3e15532f95f8f8933R426 ,  i.e. an MSI based interrupt injection through KVM? Is this to not add `VmFd` dependency on the IOAPIC implementation?

Reply:
As you mentioned, the reason for an abstraction here (callback or trait) is to prevent the IOAPIC code from being aware about the `VmFd`, which is KVM specific.